<div style="border:solid green 2px; padding: 20px"> 
    
Hello Catharine, 
    
    
my name is Aleksandra. I'm going to review your project! Nice to meet you! 🙌

You can find my comments under the heading **«Review»**. I will categorize my comments in green, blue or red boxes like this:

<div class="alert alert-success">
    <b>Success:</b> if everything is done successfully
</div>
<div class="alert alert-warning">
    <b>Remarks:</b> if I can give some recommendations or ways to improve the project
</div>
<div class="alert alert-danger">
    <b>Needs fixing:</b> if the block requires some corrections. Work can't be accepted with the red comments
</div>

Please don't remove my comments :) If you have any questions don't hesitate to respond to my comments in a different section. 

<div class="alert alert-block alert-success">
<b>Overall reviewer's comment v1</b> <a class="tocSkip"></a>
    
Catharine, thank you so much for the project! I can see that a lot of effort has been made and it turned out to be very strong. You work very carefully with data and make qualitative assumptions - like a real analyst.
    
There are no critical comments. I left a few directional comments. If you have any questions, feel free to ask them through your tutor. Good luck in the next steps! 
</div>

# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.

# Project Purpose
The purpose of this project is to determine if a customer's marital status and number of children correlates to their likelihood to default on a loan. The conclusion will be used to inform how to build credit scores for potential customers. 

# Hypothesis
A customer's marital status and the number of children will impact their likelihood to repay a loan. Specifically:
* Being unmarried will correlate with a greater likelihood of defaulting on a loan.
* Having children will correlate with a greater likelihood of defaulting on a loan.

## Open the data file and have a look at the general information. 

In [372]:
import pandas as pd # Loading all the libraries
import math

loan_data = pd.read_csv('/datasets/credit_scoring_eng.csv') # Load the data


## Task 1. Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan

In [373]:
loan_data.shape # Checks the number of rows and columns our dataset has

(21525, 12)

In [374]:
loan_data.head(10) # Prints the first 10 rows

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

**Observations**:

The following are initial observations that need to be further investigated:
* Days_employed column has negative numbers
* Age seems to concentrated around 30s and 40s
* Education column has inconsistent use of lower and capital letters
* Purpose column has inconsistent language across similar categories 
* Do not see any missing values in this initial snapshot, but needs further analysis

<div class="alert alert-success">
<b>Reviewer's comment v1</b> Correct

In [375]:
loan_data.info() # Gets general info on data

loan_data.isna().sum() # Looks at number of missing values in each column

loan_data.isna().sum()/len(loan_data) * 100 # Looks at percentage of missing values compared to the whole dataset 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children             0.000000
days_employed       10.099884
dob_years            0.000000
education            0.000000
education_id         0.000000
family_status        0.000000
family_status_id     0.000000
gender               0.000000
income_type          0.000000
debt                 0.000000
total_income        10.099884
purpose              0.000000
dtype: float64

**Observations**:

The 'days_employed" and "total_income" columns have missing values. They contain the same number of missing values (2174), which make up 10% of the total data. 

In [376]:
loan_data.loc[loan_data['days_employed'].isna()]# Looks at a filtered table with missing values in the 'days_employed' column

loan_data.loc[loan_data['total_income'].isna()] # Looks at filtered table with missing values in the total_income column to compare to the table above. 
 
loan_data['days_employed'].isna().sum() # Looks at the total number of null values in the column

loan_data['total_income'].isna().sum() # Looks at the total number of null values in the column


2174

**Observations:**

The 'days_employed' and the 'total_income' columns have the same number of missing datapoints, which may mean that rows with missing days of employment are also missing income. We will need to further investigate to confirm that this is true.   


In [377]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.

table_filtered = loan_data.loc[loan_data['total_income'].isna() & loan_data['total_income'].isna()] # Creates a filtered table that contains only the rows where 'total_income' and 'days_employed' have null values

table_filtered.shape # Checks number of rows and columns in the filtered table

(2174, 12)

**Intermediate conclusion**

* The number of rows in the filtered table equals the number of missing values in each column.
* All rows that have a missing value in days_employed also have a missing value in total_income and vice versa.

Per my calculations above, the missing values account for about 10% of the dataset. As a next step, I will look at the filtered table and analyze whether the null values are random or if they coorelate with other categories. 

<div class="alert alert-success">
<b>Reviewer's comment v1</b> Well done

In [378]:
table_filtered # Calls this filtered table to look at client characteristics only for rows with missing data. 

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

**Observations:**

In this initial view (needs further analysis):
* Most of the clients do not have children. 
* Clients skew older (50s and 60s)
* Most have a secondary education
* There are no observable differences in family_status, gender, income_type or purpose. 
* Most did not have debt on loan repayment.



In [379]:
# Checking distribution

def distrib(col): # Defines a function to look at distribution across different columns.
    column2 = table_filtered[col]
    return table_filtered.groupby(column2)['education'].count() / len(table_filtered) * 100 

distrib('children') # Looks at distribution of missing values across number of children

distrib('dob_years') # Looks at distribution of missing values across age

distrib('education_id') #  Looks at distribution of missing values across type of education. Used education_id due to duplicates in the education column.

distrib('family_status_id') # Looks at distribution of missing values across family status. Used family_status_id due to duplicates in the family status column.

distrib('gender') # Looks at distribution of missing values across gender

distrib('income_type') # Looks at distribution of missing values across income type

distrib('debt') # Looks at distribution of missing values across debt

distrib('purpose') # Looks at distribution of missing values across purpose



purpose
building a property                         2.713891
building a real estate                      2.115915
buy commercial real estate                  3.081877
buy real estate                             3.311868
buy residential real estate                 2.805888
buying a second-hand car                    1.931923
buying my own car                           2.437902
buying property for renting out             2.989880
car                                         1.885925
car purchase                                1.977921
cars                                        2.621895
construction of own property                3.449862
education                                   1.931923
getting an education                        2.299908
getting higher education                    1.655934
going to university                         2.575897
having a wedding                            4.231831
housing                                     2.759890
housing renovation                    

**Observations:**

* Clients without any children are more likely to have missing days employed/income: 66% of the clients with missing days employed/income had 0 children.
* Missing values are fairly evenly distributed across age, although slightly more concentrated in mid-20s to mid-60s. 
* Clients with secondary education are most likely to have missing values in days employed/income (70%) followed by those with a bachelor's degree(25%). 
* Clients who are married are most likely to have missing days employed/income (57%), followed up those in a civil partnershop (20%).
* Female clients are more likely to have missing data (68%)
* Employees are more likely to have missing data (51%), followed by business (23%) as income type and then retiree (19%)
* Clients with missing values were much more likely to have no debt on loan repayment (92%).
* There did not seem to be a relationship between loan purpose and missing values.

**Possible reasons for missing values in data:**
* My hypothesis is that the missing values are distributed throughout the original dataframe at random, meaning that the same characteristics observed above will be also be reflected in the entire dataset. 



In [380]:
# Checking the distribution in the whole dataset

def distrib(col): # Defines a function to look at distribution across different columns.
    column2 = loan_data[col]
    return loan_data.groupby(column2)['education'].count() / len(loan_data) * 100 

distrib('children') # Looks at distribution across number of children

distrib('dob_years') # Looks at distribution  across age

distrib('education_id') #  Looks at distribution across type of education. Used education_id due to duplicates in the education column.

distrib('family_status_id') # Looks at distribution across family status. Used family_status_id due to duplicates in the family status column.

distrib('gender') # Looks at distribution across gender

distrib('income_type') # Looks at distribution across income type

distrib('debt') # Looks at distribution across debt

distrib('purpose') # Looks at distribution across purpose

purpose
building a property                         2.880372
building a real estate                      2.908246
buy commercial real estate                  3.084785
buy real estate                             2.898955
buy residential real estate                 2.819977
buying a second-hand car                    2.225319
buying my own car                           2.346109
buying property for renting out             3.033682
car                                         2.299652
car purchase                                2.146341
cars                                        2.220674
construction of own property                2.950058
education                                   2.076655
getting an education                        2.058072
getting higher education                    1.979094
going to university                         2.304297
having a wedding                            3.609756
housing                                     3.005807
housing renovation                    

**Conclusions**

The distribution of factors (ex. number of children, age, employment status, etc.) within the filtered table that represents only rows with missing values, is similar to the larger dataset. This means that my earlier hypothesis was correct, and that the missing values are likely randomly distributed throughout the data. 

Based on the analysis above, we can determine that the missing values are distributed throughout the dataset at random. I did not find any patterns that were unique to the rows with missing values, compared to the larger table. I looked at the distribution of missing values across all of the client characteristics (ex. number of children, age, employment status, etc.) and found that the distribution was similar in the overall dataset compared to only the rows with missing values. Therefore, I concluded at the missing values are present at random throughout the table. 

Both of the columns that have missing data are represented by numbers (float type). Therefore I will take the following steps to fill in the missing values:
* Look at the relationship between days employed and total income with other categories - ex. age, family status and education. 
* Based on this analysis, I will calculate the mean (if there are not extreme outliers) or median of days employed/total income according to the selected category, and replace the missing value with that number. 

In addition to the missing values, I plan to address the following issues: 
* Columns that contain negative numbers (ex. The number of children and days_employed columns)
* The inconsistent use of lower and capital letters (ex. The education column)
* Inconsistent language/dulplicates across similar categories (ex. The purpose column)

<div class="alert alert-success">
<b>Reviewer's comment v1</b> We could report bank about the problem and ask it for meta data for further investigation

## Data transformation

**Fixing the duplicate values in the education column**

In [381]:
loan_data['education'].sort_values().unique() # Checks for unique values in the education column, sorted alphabetically so that it's easier to read


array(["BACHELOR'S DEGREE", "Bachelor's Degree", 'GRADUATE DEGREE',
       'Graduate Degree', 'PRIMARY EDUCATION', 'Primary Education',
       'SECONDARY EDUCATION', 'SOME COLLEGE', 'Secondary Education',
       'Some College', "bachelor's degree", 'graduate degree',
       'primary education', 'secondary education', 'some college'],
      dtype=object)

In [382]:
loan_data['education'] = loan_data['education'].str.lower() # Changes all of the letters to lowercase so that the column can be de-duplicated

<div class="alert alert-success">
<b>Reviewer's comment v1</b> Great! We have dealed with duplicates

In [383]:
loan_data['education'].sort_values().unique() # Checks to make sure the unique values are reflecting the change to all lowercase letters

array(["bachelor's degree", 'graduate degree', 'primary education',
       'secondary education', 'some college'], dtype=object)

**Fixing the values in the children column**

In [384]:
loan_data.groupby('children')['education'].count() # Looks at distribution of values across children's ages

loan_data.groupby('children')['education'].count() / len(loan_data) * 100 # Looks at distribution of values by percentage across children's ages

children
-1      0.218351
 0     65.732869
 1     22.383275
 2      9.547038
 3      1.533101
 4      0.190476
 5      0.041812
 20     0.353078
Name: education, dtype: float64

**Observations**

* 47 rows have a negative number for the number of children which isn't possible.
* This only represents .2% of the data. 
* This mistake is most likely due to entry error. It could be that these families are supposed to have 0 children, or it could be families with 1 child that accidentally got coded as a negative number. Since there is no way for us to know, and because these data represent such a small percentage of the overall dataset and will not impact our overall analysis, I will choose to drop these rows. 

<div class="alert alert-success">
<b>Reviewer's comment v1</b> Great that you have found that abnormal value! Now the proportion of anomalies is small and it will not affect anything

In [385]:
children_neg = loan_data.loc[loan_data['children'] == -1].index
                             
loan_data.drop(children_neg, inplace=True) # Drops the rows where the number of children is -1

In [386]:
loan_data.groupby('children')['education'].count() # Checks the distribution of values across children's ages to make sure we removed -1

loan_data.groupby('children')['education'].count() / len(loan_data) * 100 # Looks at distribution of values by percentage across children's ages. This also shows me that the percentage of clients with no children did not change signficantly.

children
0     65.876711
1     22.432256
2      9.567930
3      1.536456
4      0.190893
5      0.041903
20     0.353850
Name: education, dtype: float64

**Fixing the values in the days_employed column**

I saw previously that this column has negative numbers. This could again be entry error and could negatively impact analysis. 

In [387]:
loan_data.loc[loan_data['days_employed'] < 0].shape # Locates all of the rows for which days_employed is a negative number, and returns the number of rows/columns. This tells us that there are 15,906 instances of a negative number.

loan_data.loc[loan_data['days_employed'] < 0]['children'].count() / len(loan_data) * 100 # Looks at percentage of rows that have a negative number for days_employed.

73.884905484682

**Observations**

* 74% of rows have a negative number for days employed, which is too high to drop from the table. 
* Due to the large amount of data this impacts, I will assume that these numbers were incorretly entered as negative and change them from a negative to a positive number so that I can use them in the analysis.

In [388]:
loan_data['days_employed'] = loan_data['days_employed'].abs() # Changes all of the numbers in the column to positive numbers. 


In [389]:
loan_data['days_employed'].sort_values() # Checks values, sorted from low to high, which shows us that there are no negative numbers.

loan_data['days_employed'].min() # Checks the minimum number in the 'days_employed' column to make sure that it's not a negative number.

24.14163324048118

**Fixing the data in the age (dob_years) column**

In [390]:
loan_data['dob_years'].sort_values().unique() # Looks at the unique values in 'dob_years', sorted from low to high.

loan_data.groupby('dob_years')['education'].count() # Looks at count of rows per age specified under "'dob_years'

loan_data.groupby('dob_years')['education'].count() /len(loan_data) * 100  # Looks at percentage of rows per age specified under 'dob_years'

loan_data.loc[loan_data['dob_years'] == 0]['children'].count() / len(loan_data) * 100  # Double checks the numbers above, this looks at percentage of rows where the 'dob_years' = 0


0.4702486265015364

**Observations**

* One of the ages specified is 0 which does not make sense. 
* From the analysis above we know that rows with the age 0 under 'dob_years' make up only .5% of total rows in the database.
* Due to the small percentage of rows that this makes up, I would choose to drop these rows since it will not impact the overall analysis. 

<div class="alert alert-success">
<b>Reviewer's comment v1</b> Well done

In [391]:
drop_age = loan_data.loc[loan_data['dob_years'] == 0].index # Identifies the rows where age = 0
                             
loan_data.drop(drop_age, inplace=True) # Drops rows where the age = 0 

In [392]:
loan_data.groupby('dob_years')['education'].count() # Checks to make sure 0 has been dropped from the data. 

dob_years
19     14
20     51
21    111
22    183
23    253
24    264
25    357
26    407
27    492
28    501
29    545
30    539
31    558
32    508
33    579
34    600
35    616
36    555
37    535
38    596
39    573
40    607
41    605
42    595
43    512
44    545
45    497
46    472
47    480
48    537
49    508
50    512
51    447
52    484
53    458
54    477
55    442
56    487
57    458
58    461
59    443
60    377
61    354
62    352
63    268
64    264
65    194
66    183
67    167
68     99
69     84
70     65
71     58
72     33
73      8
74      6
75      1
Name: education, dtype: int64

**Examining the data in the family status column**

In [393]:
loan_data['family_status'].sort_values().unique() # Looks at the unique values, sorted alphabetically

loan_data.groupby('family_status')['education'].count() /len(loan_data) * 100  # Looks at percentage of rows per category specified under 'family_status', just to get a snapshot of the different categories


family_status
civil partnership    19.418066
divorced              5.524629
married              57.547832
unmarried            13.060766
widow / widower       4.448707
Name: education, dtype: float64

**Observations:**
* I do not see any obvious problems - no duplicate categories and no syntax issues.
* You could argue that the civil partnership and marriage represent a similar category around long-term partnership and could be combined. Since the bank collected these categories separately I will assume that they want to see the analysis for each category, and will leave them separate. 

**Examining the data in the gender column**

In [394]:
loan_data['gender'].sort_values().unique() # Looks at the unique values, sorted alphabetically

loan_data.groupby('gender')['education'].count() # Looks at count of each gender category.

loan_data.groupby('gender')['education'].count() /len(loan_data) * 100  # Looks at percentage of rows per gender category


gender
F      66.094401
M      33.900922
XNA     0.004678
Name: education, dtype: float64

**Observations**
* There is one row that contains the gender category "XNA". This could be a coding error, could represent a missing gender value, or could represent it's own gender category (ex. non-binary). 
* Regardless of whether this row represent it's own gender category or an error, it is not enough data to do analysis with. I will change the "XNA" to "Unknown". 

In [395]:
loan_data['gender'] = loan_data['gender'].replace('XNA','Unknown') # Replaced the gender XNA with Unknown

<div class="alert alert-success">
<b>Reviewer's comment v1</b> Good decision!

In [396]:
loan_data.groupby('gender')['education'].count() # Looks at count of each gender category.

loan_data.groupby('gender')['education'].count() /len(loan_data) * 100  # Looks at percentage of rows per gender category


gender
F          66.094401
M          33.900922
Unknown     0.004678
Name: education, dtype: float64

**Examining the income_type column**

In [397]:
loan_data['income_type'].sort_values().unique() # Looks at the unique values, sorted alphabetically

loan_data.groupby('income_type')['education'].count() # Looks at count of each income type category.

loan_data.groupby('income_type')['education'].count() /len(loan_data) * 100  # Looks at percentage of rows per income type category

income_type
business                       23.651588
civil servant                   6.778313
employee                       51.634935
entrepreneur                    0.009356
paternity / maternity leave     0.004678
retiree                        17.907096
student                         0.004678
unemployed                      0.009356
Name: education, dtype: float64

**Observations:**
* I do not see any obvious problems with these categories. 
* I do think that it would be helpful to understand the difference between "employee" and other categories (ex. since a "civil servant" is also an "employee", but without knowing more details I will assume that the bank wants to see the difference between these two categories. 

<div class="alert alert-success">
<b>Reviewer's comment v1</b> Correct

**Checking for duplicates**

In [398]:
loan_data.info() # Prints the basic table information, which I used to remind myself of all of the column names/information

loan_data.duplicated().sum() # Returns the number of duplicate rows (71)

loan_data.duplicated().sum() / len(loan_data) * 100 # Returns the percentage of duplicate rows compared to the entire dataframe 

loan_data[loan_data.duplicated()] # Returns the rows that are duplicates, allows me to get a snapshot of these duplicate rows


<class 'pandas.core.frame.DataFrame'>
Int64Index: 21377 entries, 0 to 21524
Data columns (total 12 columns):
children            21377 non-null int64
days_employed       19216 non-null float64
dob_years           21377 non-null int64
education           21377 non-null object
education_id        21377 non-null int64
family_status       21377 non-null object
family_status_id    21377 non-null int64
gender              21377 non-null object
income_type         21377 non-null object
debt                21377 non-null int64
total_income        19216 non-null float64
purpose             21377 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


children  days_employed  dob_years            education  education_id  \
2849          0            NaN         41  secondary education             1   
3290          0            NaN         58  secondary education             1   
4182          1            NaN         34    bachelor's degree             0   
4851          0            NaN         60  secondary education             1   
5557          0            NaN         58  secondary education             1   
...         ...            ...        ...                  ...           ...   
20702         0            NaN         64  secondary education             1   
21032         0            NaN         60  secondary education             1   
21132         0            NaN         47  secondary education             1   
21281         1            NaN         30    bachelor's degree             0   
21415         0            NaN         54  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
2849             married                 0      F    employee     0   
3290   civil partnership                 1      F     retiree     0   
4182   civil partnership                 1      F    employee     0   
4851   civil partnership                 1      F     retiree     0   
5557   civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
20702            married                 0      F     retiree     0   
21032            married                 0      F     retiree     0   
21132            married                 0      F    employee     0   
21281            married                 0      F    employee     0   
21415            married                 0      F     retiree     0   

       total_income                              purpose  
2849            NaN  purchase of the house for my family  
3290            NaN                    to have a wedding  
4182            NaN                     wedding ceremony  
4851            NaN                     wedding ceremony  
5557            NaN                    to have a wedding  
...             ...                                  ...  
20702           NaN              supplementary education  
21032           NaN                   to become educated  
21132           NaN                   housing renovation  
21281           NaN           buy commercial real estate  
21415           NaN                 housing transactions  

[71 rows x 12 columns]

**Observations:**
* There are 71 duplicate rows, which make up .33% of the entire dataframe
* I could choose to drop the duplicate rows using the .dropduplicates() function, but looking back at the basic table information, there are no columns that uniquely identify the individuals (ex. UserID, social security number, license number, etc.). Therefore, it is possible that two individuals could have the same information and we cannot assume that duplicates are belong to the same individual. 

In [399]:
loan_data.shape # Checks the new size of the dataset

(21377, 12)

**Observations**
* In summary, I chose to remove rows for which there were problematic values, which were rows where clients were indicated to have "-1" children, and clients with an age of 0. 
* I chose not to remove duplicate rows, since there are no unique individual identifiers and technically clients could share the same demographic and loan informaton. 
* In the end, this resulted in removiing a total of 148 rows (21,525 rows to 21,377 rows), which represent 0.7% of the original number of rows. 

<div class="alert alert-success">
<b>Reviewer's comment v1</b> We have found all the duplicates. Great!

# Working with missing values

I created two dictionaries below - one to define reference IDs for family status and the other to define reference Ids for level of education. I want to be able to take both of these factors into account in my analysis below, and therefore want to use these dictionaries as reference in that analysis. 

In [400]:
family_status_dict = loan_data[['family_status_id', 'family_status']].drop_duplicates().reset_index(drop=True) # Creates a family status dictionary

education_dict = loan_data[['education_id', 'education']].drop_duplicates().reset_index(drop=True) # Creates an education_id dictionary

family_status_dict

education_dict


education_id            education
0             0    bachelor's degree
1             1  secondary education
2             2         some college
3             3    primary education
4             4      graduate degree

### Restoring missing values in `total_income`

**Initial observations and plans for missing values:**

Two columns in this dataset have missing values: days_employed and total_income. Both have the float type. Therefore I will take the following steps to fill in the missing values:
* Look at the relationship between days employed and total income with other categories - ex. number of children, age, etc. 
* Based on this analysis, I will calculate the mean (if there are not extreme outliers) or median (if there are extreme outliers that impact the mean) of days employed/total income according to the selected category, and replace the missing value with that number.


In [401]:
# Let's write a function that calculates the age category
def age_group(row): # Function that creates age groups according to decade which is common practice
    age = row['dob_years']
    if age < 10:
        return '0-9'
    elif age < 20:
        return '10-19'
    elif age < 30:
        return '20-29'
    elif age < 40:
        return '30-39'
    elif age < 50:
        return '40-49'
    elif age < 60:
        return '50-59'
    elif age < 70:
        return '60-69'
    else:
        return '70+'

In [402]:
row_values = [47] # Creates a list that indicates row value
row_columns = ['dob_years'] # Creates list that indicates column name

row = pd.Series(data=row_values, index=row_columns) # Creates a new row with the above values

print(age_group(row)) # Passes the row through the function

40-49


In [403]:
loan_data['age_group'] = loan_data.apply(age_group, axis=1) # Applys the function to loan_data and returns the result in a new column called 'age_group'

In [404]:
loan_data.head(10) # Prints the first 10 rows to confirm the new column has been added

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose age_group  
0                purchase of the house     40-49  
1                         car purchase     30-39  
2                purchase of the house     30-39  
3              supplementary education     30-39  
4                    to have a wedding     50-59  
5                purchase of the house     20-29  
6                 housing transactions     40-49  
7                            education     50-59  
8                     having a wedding     30-39  
9  purchase of the house for my family     40-49

In [405]:
loan_data_filtered = loan_data.loc[loan_data['total_income'] > 0] # Creates a table without missing values

loan_data_filtered.head(10) # Prints first 10 rows


children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose age_group  
0                purchase of the house     40-49  
1                         car purchase     30-39  
2                purchase of the house     30-39  
3              supplementary education     30-39  
4                    to have a wedding     50-59  
5                purchase of the house     20-29  
6                 housing transactions     40-49  
7                            education     50-59  
8                     having a wedding     30-39  
9  purchase of the house for my family     40-49

In [406]:
loan_data_filtered.groupby('age_group')['total_income'].mean() # Finds the mean of the income column for each age group


age_group
10-19    16993.942462
20-29    25575.666766
30-39    28316.552125
40-49    28565.424392
50-59    25821.596685
60-69    23228.696417
70+      20125.658331
Name: total_income, dtype: float64

In [407]:
loan_data_filtered.groupby('age_group')['total_income'].median() # Finds the median of the income column for each age group

age_group
10-19    14934.9010
20-29    22801.8085
30-39    24692.4500
40-49    24768.1370
50-59    22201.4960
60-69    19811.3260
70+      18751.3240
Name: total_income, dtype: float64

In [408]:
loan_data_filtered.groupby('age_group').agg({'total_income': [max, min]}) #Looks at range of data  across each age group

total_income          
                   max       min
age_group                       
10-19        26753.823  9459.851
20-29       131588.163  4494.861
30-39       352136.354  3392.845
40-49       362496.645  4036.463
50-59       195686.797  3306.762
60-69       274402.943  3471.216
70+          57508.032  3595.641

In [409]:
income_median = loan_data_filtered.pivot_table(index=['age_group'], columns=['education_id'], values='total_income', aggfunc='median')

income_median


education_id           0           1           2           3           4
age_group                                                               
10-19                NaN  15761.5690  14575.7170         NaN         NaN
20-29         25948.0270  21176.7800  22698.1735  25488.9160         NaN
30-39         28801.6055  22907.8415  28463.4390  19542.3265  18187.3015
40-49         30322.0090  22971.2210  29323.6730  21511.5635  31771.3210
50-59         28153.0340  21236.0770  21132.3090  16922.6250  42945.7940
60-69         25195.1730  18873.7640  28058.6760  17657.4995  28334.2150
70+           25497.3920  18508.5770  14479.1930  15013.5050         NaN

**Observations**

* Looking at the breakdown across age groups, the mean is larger than the median, which indicates that there are outliers. This was confirmed by looking at the minimum and maximum income in each age category. 
* Therefore, the median will be a better representation of the income data. 
* Age group is not the only category that can influence income. I looked at the breakdown of income by education level and family status for each age group. Income was not greatly impact by family status, but it was impacted by education level. Therefore I will use the median of age group and education level to fill in missing values. 

<div class="alert alert-success">
<b>Reviewer's comment v1</b> Great!

In [410]:
def fill_income(row): # Creates a function to fill in missing values in the total_income column
    age_group = row['age_group']
    education_id = row['education_id']
    total_income = row['total_income']
    if math.isnan(total_income):
        return income_median[education_id][age_group]
    else:
        return total_income

In [411]:
row_values = ['70+', 0, float('NaN')] # Creates a row
row_columns = ['age_group', 'education_id', 'total_income'] # Creates columns
row = pd.Series(data=row_values, index=row_columns) # Passes the row and columns through the function to make sure we get the expected result

fill_income(row)


25497.392000000003

In [412]:
loan_data['total_income_new'] = loan_data.apply(lambda x: fill_income(x), axis=1) # Applies the function to the loan_data dataframe and returns value to new column

loan_data.head(30) # Checks that the table has the new column 

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42    bachelor's degree             0   
1          1    4024.803754         36  secondary education             1   
2          0    5623.422610         33  secondary education             1   
3          3    4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0     926.185831         27    bachelor's degree             0   
6          0    2879.202052         43    bachelor's degree             0   
7          0     152.779569         50  secondary education             1   
8          2    6929.865299         35    bachelor's degree             0   
9          0    2188.756445         41  secondary education             1   
10         2    4171.483647         36    bachelor's degree             0   
11         0     792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0    1846.641941         54         some college             2   
14         0    1844.956182         56    bachelor's degree             0   
15         1     972.364419         26  secondary education             1   
16         0    1719.934226         35  secondary education             1   
17         0    2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0   10038.818549         48  secondary education             1   
20         1    1311.604166         36  secondary education             1   
21         1     253.685166         33  secondary education             1   
22         1    1766.644138         24  secondary education             1   
23         0     272.981385         21    bachelor's degree             0   
24         1  338551.952911         57  secondary education             1   
25         0  363548.489348         67  secondary education             1   
26         0            NaN         41  secondary education             1   
27         0     529.191635         28    bachelor's degree             0   
28         1     717.274324         26    bachelor's degree             0   
29         0            NaN         63  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
0             married                 0      F       employee     0   
1             married                 0      F       employee     0   
2             married                 0      M       employee     0   
3             married                 0      M       employee     0   
4   civil partnership                 1      F        retiree     0   
5   civil partnership                 1      M       business     0   
6             married                 0      F       business     0   
7             married                 0      M       employee     0   
8   civil partnership                 1      F       employee     0   
9             married                 0      M       employee     0   
10            married                 0      M       business     0   
11            married                 0      F       employee     0   
12  civil partnership                 1      M        retiree     0   
13            married                 0      F       employee     0   
14  civil partnership                 1      F       business     1   
15            married                 0      F       employee     0   
16            married                 0      F       employee     0   
17  civil partnership                 1      M       employee     0   
18    widow / widower                 2      F        retiree     0   
19           divorced                 3      F       employee     0   
20            married                 0      M       employee     0   
21  civil partnership                 1      F       e

In [413]:
loan_data['total_income_new'].isna().sum() # Checks to make sure the new column got rid of null values 

0

In [414]:
loan_data.info() # Checks the number of entries in the columns

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21377 entries, 0 to 21524
Data columns (total 14 columns):
children            21377 non-null int64
days_employed       19216 non-null float64
dob_years           21377 non-null int64
education           21377 non-null object
education_id        21377 non-null int64
family_status       21377 non-null object
family_status_id    21377 non-null int64
gender              21377 non-null object
income_type         21377 non-null object
debt                21377 non-null int64
total_income        19216 non-null float64
purpose             21377 non-null object
age_group           21377 non-null object
total_income_new    21377 non-null float64
dtypes: float64(3), int64(5), object(6)
memory usage: 2.4+ MB


###  Restoring values in `days_employed`

In [415]:
days_employed_median = loan_data_filtered.pivot_table(index=['age_group'], columns=['education_id'], values='days_employed', aggfunc='median') # Looks at distribution of `days_employed` medians based on education_id

days_employed_median


education_id              0              1              2              3  \
age_group                                                                  
10-19                   NaN     640.290559     796.983636            NaN   
20-29           1005.389899    1018.350987     891.699189    1030.219648   
30-39           1598.195256    1639.634599    1209.230373    1110.847169   
40-49           2080.471069    2130.417977    2063.961740    1108.837429   
50-59           4109.333294    5131.992936    2917.701025  334483.493054   
60-69         342046.496714  356180.806455  352184.250110  358139.948376   
70+           352635.085407  362034.079738  396422.414465  360639.907161   

education_id              4  
age_group                    
10-19                   NaN  
20-29                   NaN  
30-39           4159.753909  
40-49            409.200149  
50-59           5352.038180  
60-69         356930.517546  
70+                     NaN

In [416]:
days_employed_mean = loan_data_filtered.pivot_table(index=['age_group'], columns=['education_id'], values='days_employed', aggfunc='mean') # Looks at distribution of `days_employed` means based on education_id

days_employed_mean

education_id              0              1              2              3  \
age_group                                                                  
10-19                   NaN     550.174786     767.283365            NaN   
20-29           1643.408086    2314.388515    1050.326009    1236.562385   
30-39           3325.186149    4623.358528    1574.576678    7233.407335   
40-49          11406.384940   12544.778139   12673.184467   17720.818483   
50-59         109021.282638  137948.961319  110861.951090  204410.044227   
60-69         234253.941889  294402.445053  234832.804041  321462.637175   
70+           303999.187231  320231.552828  381900.045298  360567.950991   

education_id              4  
age_group                    
10-19                   NaN  
20-29                   NaN  
30-39           4159.753909  
40-49            409.200149  
50-59           5352.038180  
60-69         356930.517546  
70+                     NaN

**Observations**

By comparing the medians and means, we can see that the means for some categories vary significantly from the medians, indicating there are outliers for those data. Therefore, I will use the medians to represent the data. 

In [417]:
def fill_days_employed(row): # Creates a function to fill in missing values in days_employed with the median based on age_group and education level 
    age_group = row['age_group']
    education_id = row['education_id']
    days_employed = row['days_employed']
    if math.isnan(days_employed):
        return days_employed_median[education_id][age_group]
    else:
        return days_employed


In [418]:
row_values = ['20-29', 1, float('NaN')] # Creates a row
row_columns = ['age_group', 'education_id', 'days_employed'] # Creates columns 
row = pd.Series(data=row_values, index=row_columns) # Passes the row and columns through the function to make sure we get the expected result

fill_days_employed(row)


1018.3509869969278

In [419]:
loan_data['days_employed_new'] = loan_data.apply(lambda x: fill_days_employed(x), axis=1) # Applys the function to the income_type



In [420]:
loan_data.head(30) # Checks that the function worked by calling the dataframe

loan_data['days_employed_new'].isna().sum() # Checks to make sure we no longer have null values in the new days employed column

0

<div class="alert alert-success">
<b>Reviewer's comment v1</b> Great job on filling missing values!

## Categorization of data

**Observations**

* Based on the data needed to test the hypothesis, I will look at the columns children, family status and total_income to look at whether I need to do further categorization for analysis. 

In [421]:
loan_data.head(20) # Prints an overview of the data so I can look at snapshot of the columns of interest: children, family_status & total_income


children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42    bachelor's degree             0   
1          1    4024.803754         36  secondary education             1   
2          0    5623.422610         33  secondary education             1   
3          3    4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0     926.185831         27    bachelor's degree             0   
6          0    2879.202052         43    bachelor's degree             0   
7          0     152.779569         50  secondary education             1   
8          2    6929.865299         35    bachelor's degree             0   
9          0    2188.756445         41  secondary education             1   
10         2    4171.483647         36    bachelor's degree             0   
11         0     792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0    1846.641941         54         some college             2   
14         0    1844.956182         56    bachelor's degree             0   
15         1     972.364419         26  secondary education             1   
16         0    1719.934226         35  secondary education             1   
17         0    2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0   10038.818549         48  secondary education             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   

    total_income                              purpose age_group  \
0      40620.102                purchase of the house     40-49   
1      17932.802                         car purchase     30-39   
2      23341.752                purchase of the house     30-39   
3      42820.568              supplementary education     30-39   
4      25378.572                    to have a wedding     50-59   
5      40922.170                purchase of the house     20-29   
6      38484.156                 housing transactions     40-49   
7      21731.829                            education     50-59   
8      15337.093                     having a wedding     30-39   
9      23108.150  purchase of the house for my family     40-49   
10     18230.959                      buy real estate     30-39   
11     12331.077           buy commercial real estate     40-49   
12           NaN                    to have a wedding     60-69   
13     20873.317   

In [422]:
loan_data['children'].unique() # Checks the unique values 

loan_data['family_status'].unique() # Checks the unique values

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

**Observations**
* Based on looking at the unique values in each column, family status does not need to be further categorized. 
* I will start by categorizing the children column so that it's a binary variable ("yes" or "no" to having children).
* Then I will take a closer look at the total_income column to decide how to categorize those data.

**Creating categories for children**

In [423]:
def children_group(row): # Function that creates age groups according to decade which is common practice
    children = row['children']
    if children == 0:
        return 'no'
    elif children > 0:
        return 'yes'

In [424]:
loan_data['children_new'] = loan_data.apply(children_group, axis=1) # Creates a column with the new categories for children

loan_data['children_new'].value_counts() # Counts the values in each category

no     14080
yes     7297
Name: children_new, dtype: int64

**Creating categories for purpose**

In [425]:
loan_data['purpose'].sort_values().unique() # Looks at the unique values in the purpose column, sorted them alphabetically

array(['building a property', 'building a real estate',
       'buy commercial real estate', 'buy real estate',
       'buy residential real estate', 'buying a second-hand car',
       'buying my own car', 'buying property for renting out', 'car',
       'car purchase', 'cars', 'construction of own property',
       'education', 'getting an education', 'getting higher education',
       'going to university', 'having a wedding', 'housing',
       'housing renovation', 'housing transactions', 'profile education',
       'property', 'purchase of a car', 'purchase of my own house',
       'purchase of the house', 'purchase of the house for my family',
       'real estate transactions', 'second-hand car purchase',
       'supplementary education', 'to become educated', 'to buy a car',
       'to get a supplementary education', 'to have a wedding',
       'to own a car', 'transactions with commercial real estate',
       'transactions with my real estate', 'university education',
       'we

In [426]:
def purpose(row): # Create a function to create new categories for the purpose column
    purpose = row['purpose']
    if 'build' in purpose or 'estate' in purpose or 'property' in purpose:
        return 'real estate'
    elif 'car' in purpose:
        return 'car'
    elif 'educ' in purpose or 'university' in purpose:
        return 'education'
    elif 'housing' in purpose or 'house' in purpose: 
        return 'housing'
    elif 'wedding' in purpose:
        return 'wedding'    

In [427]:
loan_data['purpose_new'] = loan_data.apply(purpose, axis=1) # Applys the function to the entire data frame

loan_data.head(10) # Prints the first 10 rows of the dataframe to make sure the new purpose column is added 

loan_data['purpose_new'].unique() # Checks unique values in purpose_new column to make sure no values were left out

array(['housing', 'car', 'education', 'wedding', 'real estate'],
      dtype=object)

**Creating categories for total_income**

In [428]:
loan_data['total_income'].unique() # Looks at unique values in the total_income column.


array([40620.102, 17932.802, 23341.752, ..., 14347.61 , 39054.888,
       13127.587])

In [429]:
loan_data['total_income'].describe() # Gets summary statistics for total_income


count     19216.000000
mean      26799.419120
std       16504.133012
min        3306.762000
25%       16493.009500
50%       23201.873500
75%       32551.144000
max      362496.645000
Name: total_income, dtype: float64

**Observations**
* Based on the statistical description of the data, I see that the data ranges from 3306 to 362,496.
* Based on the mean of 26,799, and the fact that 75% of the data is under 32,000, I expect the data to skew low.
* Therefore, I plan to make the categories in 25,000 increments, stopping at a 100,000+ category.   

In [430]:
def total_income_func(row): # Function that creates age groups according to decade which is common practice
    total_income = row['total_income']
    if total_income < 25000:
        return '0 - 24,999'
    elif total_income < 50000:
        return '25,000 - 49,999'
    elif total_income < 75000:
        return '50,000 - 74,999'
    elif total_income < 100000:
        return '75,000 - 99,999'
    else:
        return '100,000+'

In [431]:
loan_data['total_income_new'] = loan_data.apply(total_income_func, axis=1)

loan_data.head(10)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose age_group total_income_new  \
0                purchase of the house     40-49  25,000 - 49,999   
1                         car purchase     30-39       0 - 24,999   
2                purchase of the house     30-39       0 - 24,999   
3              supplementary education     30-39  25,000 - 49,999   
4                    to have a wedding     50-59  25,000 - 49,999   
5                purchase of the house     20-29  25,000 - 49,999   
6                 housing transactions     40-49  25,000 - 49,999   
7                            education     50-59       0 - 24,999   
8                     having a wedding     30-39       0 - 24,999   
9  purchase of the house for my family     40-49       0 - 24,999   

   days_employed_new children_new purpose_new  
0        8437.673028          yes     housing  
1        4024.803754          yes         car  
2        5623.422610           no     housing  
3        4124.747207          yes   education  
4      340266.072047           no     wedding  
5         926.185831           no     housing  
6        2879.202052           no     housing  
7         152.779569           no   education  
8        6929.865299          yes     wedding  
9        2188.756445           no     housing

In [432]:
loan_data['total_income_new'].value_counts() # Counts each categories values to see the distribution

0 - 24,999         10674
25,000 - 49,999     7226
100,000+            2260
50,000 - 74,999     1028
75,000 - 99,999      189
Name: total_income_new, dtype: int64

<div class="alert alert-success">
<b>Reviewer's comment v1</b> Categorization will help to answer the questions below

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [433]:
loan_data.groupby('children_new')['debt'].sum() / loan_data['debt'].sum() * 100 # Calculates the rate of loan defaults based on whether or not families have kids


children_new
no     61.08545
yes    38.91455
Name: debt, dtype: float64

**Conclusion**

Yes, there is a correlation between having children and paying back a loan on time. Clients who do not have children are more likely to default on their loan. 


<div class="alert alert-success">
<b>Reviewer's comment v1</b> Correct

**Is there a correlation between family status and paying back on time?**

In [434]:
loan_data.groupby('family_status')['debt'].sum() / loan_data['debt'].sum() * 100 # Calculates the rate of loan defaults based on family status


family_status
civil partnership    22.286374
divorced              4.907621
married              53.464203
unmarried            15.762125
widow / widower       3.579677
Name: debt, dtype: float64

**Conclusion**

Yes, family status has a correlation with paying back a loan on time. Being married is associated with a higher likelihood of having debt on their loan, followed by civil partnership.

**Is there a correlation between income level and paying back on time?**

In [435]:
loan_data.groupby('total_income_new')['debt'].sum() / loan_data['debt'].sum() * 100 # Calculates the rate of loan defaults based on income categories


total_income_new
0 - 24,999         51.385681
100,000+           10.161663
25,000 - 49,999    33.487298
50,000 - 74,999     4.387991
75,000 - 99,999     0.577367
Name: debt, dtype: float64

**Conclusion**

Yes, income has a correlation with paying back a loan on time. A lower income is associated with a higher likelihood of having debt on their loan.

**How does credit purpose affect the default rate?**

In [436]:
# Check the percentages for default rate for each credit purpose and analyze them

loan_data.groupby('purpose_new')['debt'].sum() / loan_data['debt'].sum() * 100


purpose_new
car            23.036952
education      21.362587
housing        14.665127
real estate    30.311778
wedding        10.623557
Name: debt, dtype: float64

**Conclusion**

* Clients who have no children are more likely to default on their loan. 
* Clients who are married are more likely to default on their loan, following by those in a civil partnership. 
* Clients with a lower income are more likely to default on their loan. 
* Clients who got a loan for real estate purposes are most likely to default on their loan, followed by those who got a loan for a car and for education purposes.


<div class="alert alert-success">
<b>Reviewer's comment v1</b> Well done!

# General Conclusion 

**Summary of Process**
In summary, before conducting the analysis I did the following:
* De-duplicated values in the education column (by changing all of the values to lowercase)
* Removed rows that indicated a negative number of children and indicated an age of 0.
* Changed all of the values in days_employed to be positive.
* Filled in missing values in the days_employed and total_income columns with the median values based on age group and education status. 

**Final analysis**

After completeing these steps, I conducted the analysis to test my hypothesis. 

My original hypotheses were incorrect.

The final conclusions were:
* Clients who have no children are more likely to default on their loan. 
* Clients who are married are more likely to default on their loan, following by those in a civil partnership. 
* Clients with a lower income are more likely to default on their loan. 
* Clients who got a loan for real estate purposes are most likely to default on their loan, followed by those who got a loan for a car and for education purposes.


<div class="alert alert-success">
<b>Reviewer's comment v1</b> Thanks a lot for the project! A very usefull information for the bank